### ISSUE TITLE
Error running exe: The 'jaraco' package is required

### ISSUE THREAD

**Issue by almarklein** [pyinstaller/pyinstaller#6564]
## Description of the issue

At wgpu-py we have a [pyinstaller hook](https://github.com/pygfx/wgpu-py/tree/main/wgpu/__pyinstaller) that also gets tested on CI. Today we got an error.

Below is (what I think) is the relevant logs. The full log is [here](https://github.com/pygfx/wgpu-py/runs/5039162684?check_suite_focus=true)
```
INFO     PyInstaller.building.api:api.py:855 Building COLLECT COLLECT-00.toc
INFO     PyInstaller.building.api:api.py:896 Building COLLECT COLLECT-00.toc completed successfully.
____________________________ test_pyi_wgpu[onefile] ____________________________

pyi_builder = <PyInstaller.utils.conftest.AppBuilder object at 0x7fbad3d32f70>

    def test_pyi_wgpu(pyi_builder):
>       pyi_builder.test_source(
            """
            import wgpu.backends.rs
        """
        )
E       Failed: Running exe /tmp/pytest-of-runner/pytest-0/test_pyi_wgpu_onefile_0/dist/test_source failed with return-code 1.

/home/runner/work/wgpu-py/wgpu-py/wgpu/__pyinstaller/test_wgpu.py:2: Failed
----------------------------- Captured stdout call -----------------------------
------- Starting build. -------
------- Build finshed, now running executable. -------
[5877]  RUNNING:  '/tmp/pytest-of-runner/pytest-0/test_pyi_wgpu_onefile_0/dist/test_source' , args:  ['./test_source']
[5879]  RUNNING:  '/tmp/pytest-of-runner/pytest-0/test_pyi_wgpu_onefile_0/dist/test_source' , args:  ['./test_source']
----------------------------- Captured stderr call -----------------------------
------- Starting build. -------
------- Build finshed, now running executable. -------
[5877] PyInstaller Bootloader 3.x
...
[5878] LOADER: Running pyi_rth_inspect.py
[5878] LOADER: Running pyi_rth_pkgres.py
Traceback (most recent call last):
  File "PyInstaller/hooks/rthooks/pyi_rth_pkgres.py", line 16, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "pkg_resources/__init__.py", line 74, in <module>
  File "pkg_resources/extern/__init__.py", line 52, in create_module
  File "pkg_resources/extern/__init__.py", line 44, in load_module
ImportError: The 'jaraco' package is required; normally this is bundled with this package so if you get this warning, consult the packager of your distribution.
[5878] Failed to execute script 'pyi_rth_pkgres' due to unhandled exception!
```

### Context information (for bug reports)

* Output of `pyinstaller --version`: 4.8
* Version of Python: 3.9.10
* Version of Pytest: 6.2.5
* Platform: Ubuntu-latest (on GH Actions)
* How you installed Python: `actions/setup-python@v2`
* Did you also try this on another platform? Does it work there? Not yet.
* I did try with `develop` branch of pyinstaller.

**Comment by rokm** [comment]
Looks like this is indeed a problem with the latest `setuptools` (60.7.0) that was released six hours ago. Thanks for the heads up!

**Comment by bwoodsend** [comment]
@rokm Are you already working on a fix? We're probably going to have to get this fixed and released pretty quickly.

**Comment by rokm** [comment]
I have a rather ugly work-around that seems to work on my linux box...

**Comment by bwoodsend** [comment]
I take it that `--copy-metadata=jaraco` doesn't do it then?

**Comment by bwoodsend** [comment]
Oh, have we got a circular dependency here? Our pkg_resources runtime hook calls `import pkg_resources` but, because the hook hasn't ran yet, it can't provide the metadata for `jaraco`?

**Comment by bwoodsend** [comment]
I'd say that's a bug in setuptools then. It inhibits external metadata providers which is supposed to be a feature of `pkg_resources` so would hit any kind of embedded Python situation. This isn't just PyInstaller being lazy and not collecting things it ought to.

**Comment by rokm** [comment]
No, `jaraco` namespace package is vendored in `pkg_resources._vendor.jaraco`. But then custom loader (or finder?) and sys.path magic are used to redirect `pkg_resources.extern.jaraco` to `pkg_resources._vendor.jaraco`. Which does not seem to play nicely with our loader, in a similar way as the source-compiled OpenCV's loader script does not... So the way I got it working is to add `pkg_resources._vendor.jaraco` to `excludedimports` (just in case, it doesn't seem to be properly collected anyway) and force all stuff from `pkg_resources._vendor.jaraco` to be collected as data files. That way, their `VendoredImporter` properly kicks in, and we don't have to deal with this mess...

**Comment by bwoodsend** [comment]
Hmm, that workaround doesn't sound as gross as I expected.

**Comment by rokm** [comment]
I'll open a PR to see what our CI thinks about it, and we can use it if we fail to come up with something better...

**Comment by bwoodsend** [comment]
I could live with that.

**Comment by bwoodsend** [comment]
I see that the `sys.metapath` `pkg_resources.extern` -> `pkg_resources._vendor` yuckiness is something they were already doing and our standard hook already accounts for.

**Comment by rokm** [comment]
Yeah, but that `hiddenimports = collect_submodules('pkg_resources._vendor')` fails to pick up vendored `jaraco` - maybe because it's a namespace package? I can explicitly list hidden imports to collect -> then the problem shifts to the missing `Lorem Ipsum.txt` data file in `jaraco/text`. So we collect that, and then it suddenly turns out we need to collect it as both `pkg_resources/_vendor/jaraco/text/Lorem Ipsum.txt` and `pkg_resources/extern/jaraco/text/Lorem Ipsum.txt`... and once we do that, we get missing-module error for `jaraco.text.yield_lines`, which is actually a function from `jaraco.text` module. Which is why I opted for the work-around, because this will probably take a while to unravel...

**Comment by bwoodsend** [comment]
Looks like all this mess is simply to remove a few little string utilities. https://github.com/pypa/setuptools/pull/3061/commits/fb4a1f79f497920c3aa5b95cdfba134cc72e1dfd#diff-e3d3d454fa3a072c9f46f8affa27513892fbc2d245e87f57a5927a7be851de05

**Comment by almarklein** [comment]
> I'll open a PR to see what our CI thinks about it, and we can use it if we fail to come up with something better...

I can make our CI use that branch to check whether the fix works for us. Just let me know.

**Comment by rokm** [comment]
Here's the work-around for [develop](https://github.com/rokm/pyinstaller/tree/setuptools-60.7-workaround) and here is for [v4](https://github.com/rokm/pyinstaller/tree/setuptools-60.7-workaround-v4) branch, if you want to test it.

Looks like we'll need to sort out our support for `setuptools`' vendored `distutils` before we can run tests with setuptools 60.7 on our CI...

**Comment by almarklein** [comment]
The fix works for us!


<img width="597" alt="image" src="https://user-images.githubusercontent.com/3015475/152245744-e49a0ad4-02f6-46d4-9e22-99189bfd4850.png">

**Comment by aliencaocao** [comment]
Just a heads up, setuptools 60.7.1 also have this problem.

**Comment by rokm** [comment]
> Just a heads up, setuptools 60.7.1 also have this problem.

While that's true, setuptools 60.7.1 dropped the unnecessary `Lorem Ipsum.txt` data file, which means we can handle this now via `hiddenimports`. With 60.7.0 we have to collect that file, but that messes up the loader logic (presumably because empty directories where source files would be imply namespace (sub)packages).

So I think I'll adjust the hook update to simply add hiddenimports and mark 60.7.0 as unsupported.

**Comment by rokm** [comment]
So it turns out that my earlier work around accidentally de-vendors the `jaraco.text` package (because it calls `collect_data_files()` on `pkg_resources._vendor.jaraco.functools`, which is not a package, the path stripping does not fully work, and we end up collecting files from `pkg_resources._vendor.jaraco` into `jaraco`). While `pkg_resources` seems happy with that, it might cause issues when the application actually makes use of (real) `jaraco.text` package.

So we'll be adding support for 60.7.1 (it's just a couple of hidden imports now), but mark 60.7.0 as unsupported, since it should be trivial to either upgrade or downgrade to a compatible version.

**Comment by almarklein** [comment]
Thanks for the quick fix and release!

**Comment by ghost** [comment]
All this is a bit over my head, but I have this problem. FWIW, cx_Freeze successfully makes an exe but screws up my relative paths. I'd really like to get pyinstaller's one file to work. What can I do about this jaraco thing? Thanks.

**Comment by rokm** [comment]
> All this is a bit over my head, but I have this problem. FWIW, cx_Freeze successfully makes an exe but screws up my relative paths. I'd really like to get pyinstaller's one file to work. What can I do about this jaraco thing? Thanks.

Are you using PyInstaller 4.9?

**Comment by ghost** [comment]
> Yes, PyInstaller 4.9. And thanks for the reply.

**Comment by rokm** [comment]
> > Yes, PyInstaller 4.9. And thanks for the reply.

Then open a new issue and provide both the build log and the full error traceback.

**Comment by sfweller** [comment]
We have a customer that is getting this error message when they are executing their own installer.
When they click on the *.exe, they get the following error:

Traceback (most recent call last):
File "Lib\site-packages\PyInstaller\hooks\rthooks\pyi_rth_pkgres.py", line 16, in <module>
File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
File "pkg_resources\__init__.py", line 74, in <module>
File "pkg_resources\extern\__init__.py", line 52, in create_module
File "pkg_resources\extern\__init__.py", line 44, in load_module
ImportError: The 'jaraco.text' package is required; normally this is bundled with this package so if you get this warning, consult the packager of your distribution.

For security reasons, they are unable to install the 'jaraco.text' module from PyPI.
However, they are using setuptools version 63.4.1 together with pyinstaller 4.8, so I wouldn't expect the above error.


Any suggestions to get past this problem?

**Comment by rokm** [comment]
> For security reasons, they are unable to install the 'jaraco.text' module from PyPI. However, they are using setuptools version 63.4.1 together with pyinstaller 4.8, so I wouldn't expect the above error.
> 
> Any suggestions to get past this problem?

Is this an anaconda environment, then?

Either way, I can't reproduce this. Neither with python.org python 3.9, and pip-installed pyinstaller 4.8 & setuptools 63.4.1, nor with anaconda-installed python 3.9 and anaconda-packaged pyinstaller 4.8 & setuptools 63.4.1.

I would expect pyinstaller 4.8 to break with setuptools 60.7.1 and 60.8.0, but it seems to work out-of-the box again with later versions of setuptools (i.e., the hidden imports that we added in pyinstaller 4.9 to work around the problem are apparently not needed anymore).

Considering that the error message is about vendored `jaraco.text` as opposed to the top-level vendored `jaraco` namespace package (as was the case with original issue), this sounds as if the vendored `jaraco` package was only partially collected for some reason. Perhaps a `--hiddenimport pkg_resources._vendor.jaraco.text` would help? Otherwise, have them check their `build/<name>/Analysis-00.toc` file and find out where `pkg_resources._vendor.jaraco` was collected from, and whether other subpackages/submodules exist in that location and are listed among the entires in that .toc file.

**Comment by sfweller** [comment]
Yes, the customer is using a anaconda environment. I will give them your recommendation to use the --hiddenimports flag and point them to some documentation on using it.

Thanks for your excellent help and advice.
